exporter/cortexexporter/DESIGN.md

```diff
@@ -0,0 +1,287 @@
+
+# **OpenTelemetry Collector Cortex Exporter Design**
+
+Authors: @huyan0, @danielbang907
+
+Date: July 30, 2020
+
+## **1. Introduction**
+
+Cortex is an open source, horizontally scalable, highly available, multi-tenant, long term storage for Prometheus. Cortex accepts data defined by the Prometheus Remote Write API. Our project is focused on developing an exporter for the OpenTelemetry Collector to a Cortex instance. 
+
+The following diagram is the architecture of Cortex. 
+
+![Cortex Archietecture](https://raw.githubusercontent.com/open-o11y/opentelemetry-collector/design-doc/exporter/cortexexporter/cortex.png)
+
+### **1.1 Remote Write API**
+
+Cortex has a Remote Write API that accepts incoming metrics. This exporter should write metrics to a remote URL in a snappy-compressed, protocol buffer encoded HTTP request defined by the Remote Write API. Each request encodes multiple Cortex TimeSeries, which are composed of a set of labels and a collection of samples. Each label contains a name-value pair of strings, and each sample contains a timestamp-value number pair.
+
+![Image of TimeSeries
+](https://raw.githubusercontent.com/open-o11y/opentelemetry-collector/design-doc/exporter/cortexexporter/timeseries.png)
+
+TimeSeries stores its metric name in its labels and does not describe metric types or start timestamps. To convert to TimeSeries data, buckets of a Histogram are broken down into individual TimeSeries with a bound label(`le`), and a similar process happens with quantiles in a Summary.
+
+
+More details of Remote Write API can be found in Prometheus [documentation](https://prometheus.io/docs/prometheus/latest/storage/#overview) and Cortex [documentation](https://cortexmetrics.io/docs/apis/).
+
+### **1.2 Gaps and Assumptions**
+
+**Gap 1:** 
+Currently, metrics from the OpenTelemetry SDKs cannot be exported to Prometheus from the collector correctly ([#1255](https://github.com/open-telemetry/opentelemetry-collector/issues/1255)). This is because the SDKs send metrics to the collector via their OTLP exporter, which exports the delta value of cumulative counters. The same issue will arise for any cumulative backend service, such as Cortex.
+
+To overcome this gap in the Collector pipeline, we had proposed 2 different solutions:
+
+1. Add a [metric aggregation processor](https://github.com/open-telemetry/opentelemetry-collector/issues/1422) to the collector pipeline to aggregate delta values into cumulative values for cumulative backends like Cortex. This solution requires users to set up a collector agent next to each SDK to make sure delta values are aggregated correctly.
+2. Require the OTLP exporters in SDKs to [send cumulative values for cumulative metric types to the Collector by default](https://github.com/open-telemetry/opentelemetry-specification/issues/731). Therefore, no aggregation of delta metric values is required in the Collector pipeline for Cortex/Prometheus to properly process the data. 

```

Comment:
We cannot require/guarantee that, remember that we may also collect data from other sources.

Reply:
In that case, would the collector support some sort of routing mechanism to ensure conversion from delta to cumulative is done correctly? Or would we rely on the backend to aggregate?

Reply:
I would leave it as TBD for the moment. You have a short term way to make it work for cumulative, will revisit once the otlp is done